### DISCUSSION TITLE
use of lambda functions when registering workers

### DISCUSSION THREAD

**Discussion by jwortelboer** [eModbus/eModbus#289] [Q&A] [answered]
Hi, i am trying to use some lambda function when calling the registerWorker function in ModbusServerRTU.
For example
   server = new ModbusServerRTU(2000,1);
   server->registerWorker(0x01, READ_INPUT_REGISTER, [&](ModbusMessage request){ /* do something*/ });
but the compiler tells me its not a matching function call.

Is there a way to do this? My goal is to create a class that uses the ModbusServerRTU inside, so i have to register class function. I could create a static function but there must be a more elegant way for this...

**Comment by Miq1** [comment, marked as answer]
You are missing the return type in your lambda, the function is defined as ``std::function<ModbusMessage(ModbusMessage msg)>`` and shall return a ``ModbusMessage`` as well.

You need to define it like
```
server->registerWorker(0x01, READ_INPUT_REGISTER, [&](ModbusMessage request){ /* do something*/ } -> ModbusMessage {});
```
This is happening because your worker function seems not to have one single return statement from which the compiler can deduce the return type.

**Reply by jwortelboer** [reply]
You are absolutely spot on!  Thanks for your help.
